drivers/samsung/displays/md_series.cr

```diff
@@ -0,0 +1,500 @@
+      volume(self[:previous_volume].as_i? || 50)
+    end
+  end
+
+  enum SPEAKERMODES
+    Internal = 0
+    External = 1
+  end
+
+  def speaker_select(mode : String, **options)
+    do_send("speaker", SPEAKERMODES.parse(mode).value, **options)
+  end
+
+  def do_poll
+    do_send("status", [] of UInt8, priority: 0)
+    power? unless self[:hard_off]?
+  end
+
+  # Enable power on (without WOL)
+  def network_standby(enable : Bool, **options)
+    state = enable ? 1 : 0
+    do_send("net_standby", state, **options)
+  end
+
+  # Eco auto power off timer
+  def auto_off_timer(enable : Bool, **options)
+    state = enable ? 1 : 0
+    do_send("eco_solution", [0x81, state], **options)
+  end
+
+  # Device auto power control (presumably signal based?)
+  def auto_power(enable : Bool, **options)
+    state = enable ? 1 : 0
+    do_send("auto_power", state, **options)
+  end
+
+  # Display control
+  METHODS = [
+    "volume",
+    "contrast",
+    "brightness",
+    "sharpness",
+    "colour",
+    "tint",
+    "red_gain",
+    "green_gain",
+    "blue_gain"
+  ]
+
+  # Macro to define methods from the array above
+  {% for name in METHODS %}
+    def {{name.id}}(val : Int32, **options)
+      val = in_range(val, 100)
+      do_send({{name.id.stringify}}, val, **options)
+    end
+  {% end %}
+
+  DEVICE_SETTINGS = [
+    :network_standby,
+    :auto_off_timer,
+    :auto_power,
+    :contrast,
+    :brightness,
+    :sharpness,
+    :colour,
+    :tint,
+    :red_gain,
+    :green_gain,
+    :blue_gain,
+]
+
+  def do_device_config
+    logger.debug { "Syncronising device state with settings" }
+    DEVICE_SETTINGS.each do |name|
+      value = setting(Int32, name)
+      # TODO: find out if these are equivalent
+      # __send__(name, value) unless value.nil?
+      do_send(name.to_s, value) unless value.nil?
+    end
+  end
+
+  # TODO: check type for broadcast is correct
+  def wake(broadcast : String? = nil)
+    mac = setting(String, :mac_address)
+    if mac
+      # config is the database model representing this device
+      wake_device(mac, broadcast)
+      info = "Wake on Lan for MAC #{mac}"
+      info += " directed to VLAN #{broadcast}" if broadcast
+      logger.debug { info }
+    else
+      logger.debug { "No MAC address provided" }
+    end
+  end
+
+  enum RESPONSESTATUS
+    Ack = 0x41 # A
+    Nak = 0x4e # N
+  end
+
+  def received(data, task)
+    hex = data.hexstring
+    logger.debug { "Samsung sent: #{hex}" }
+    data = data.map{ |b| b.to_i }.to_a
+
+    # Calculate checksum of response
+    checksum = data[1..-2].sum & 0xFF
+
+    # Pop also removes the checksum from the response here
+    if data.pop != checksum
+      logger.error { "Invalid checksum\nChecksum should be: #{checksum.to_s(16)}" }
+      return task.try &.retry
+    end
+
+    status = RESPONSESTATUS.new(data[4])
+    command = COMMAND.new(data[5])
+    values = data[6..-1]
+    value = values.first
+
+    case status
+    when RESPONSESTATUS::Ack
```

Comment:
Really neat thing you get from enums: nice methods for query if a value is a specific type. In this case you can:
```crystal
case status
when .ack?
  # ...
```

Reply:
nice, thought there would be a better way to do this